src/job_runner.cc

```diff
@@ -54,6 +58,68 @@ class JobRunnerWorker {
     static constexpr std::size_t kInitialQueueCapacity = 8192;
 };
 
+class JobRunnerStrand : public std::enable_shared_from_this<JobRunnerStrand> {
+   public:
+    using job_t       = JobRunner::job_t;
+    using job_queue_t = boost::lockfree::queue<job_t*>;
+
+    JobRunnerStrand(std::weak_ptr<JobRunner> runner) : runner_weak_(runner) {}
+
+    bool AddJob(job_t&& job);
+
+    std::weak_ptr<JobRunner> runner_weak_;
+    SpinMutex                spin_mtx_;
+    bool                     has_ready_job_{false};
+    job_queue_t              pending_jobs_{kInitialQueueCapacity};
+
+    static constexpr std::size_t kInitialQueueCapacity = 8192;
+};
+
+bool JobRunnerStrand::AddJob(JobRunnerStrand::job_t&& job) {
+    auto runner = runner_weak_.lock();
+    if (!runner) {
+        LOG(WARNING) << __func__ << ": Not bound to an active runner.";
+        return false;
+    }
+
+    auto serialized_job = [job = std::move(job), strand_weak = weak_from_this()] {
+        job();
+
+        auto strand = strand_weak.lock();
+        if (!strand) {
+            return;
+        }
+
+        std::unique_lock lock(strand->spin_mtx_);
```

Comment:
Why do we use spin lock? Usually we only need at most hybrid lock, as spinning waste CPU resources and make things worse in congestion. If it's not for low latency task (e.g. rt/ddl scheduling), spinlock is almost always bad.